src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -9785,9 +9838,22 @@ private void VisitThisOrBaseReference(BoundExpression node)
                 TrackNullableStateForAssignment(right, leftLValueType, MakeSlot(left), rightState, MakeSlot(right));
             }
 
+            ForceDynamicTypeForAssignmentResultIfNecessary(node, left);
```

Comment:
My suggestion to call `SetResult` or `ForceDynamicResultType` wouldn't work too easily here :-/